### PR TITLE
Virtual Memory Manager

### PR DESCRIPTION
**Pull request by lbeziaud**

### CODE REVIEW

File: nachos/kernel/scheduler.cc [hunk 1]
```diff
@@ -124,10 +114,18 @@ void Scheduler::SwitchTo(Thread *nextThread) {
 	// we need to delete its carcass.  Note we cannot delete the thread
 	// before now (for example, in Thread::Finish()), because up to this
 	// point, we were still running on the old thread's stack!
-	if (g_thread_to_be_destroyed == oldThread) {
-		g_thread_to_be_destroyed = NULL;
-		delete oldThread;
+
+#ifdef ETUDIANTS_TP
```
Comment (sbihel) [inline nachos/kernel/scheduler.cc]: threads aren't deleted?

File: nachos/kernel/thread.h [hunk 2]
```diff
@@ -127,6 +130,8 @@ class Thread {
   //! Thread context
   threadContextT thread_context;
 
+  int local_i_clock;
```
Comment (sbihel) [inline nachos/kernel/thread.h]: I think this is quite heavy and dangerous as it can be modified from the exterior

File: nachos/kernel/thread.cc [hunk 3]
```diff
@@ -267,9 +273,12 @@ void Thread::Finish() {
 	DEBUG('t', (char *)"Finishing thread \"%s\"\n", GetName());
 	IntStatus oldStatus = g_machine->interrupt->GetStatus();
 	g_machine->interrupt->SetStatus(INTERRUPTS_OFF);
-	g_thread_to_be_destroyed = this;
-	g_alive->RemoveItem(this);
+
+  g_thread_to_be_destroyed = this;
```
Comment (sbihel) [inline nachos/kernel/thread.cc]: did you just messed the indentation?

File: nachos/vm/physMem.cc [hunk 4]
```diff
@@ -197,24 +181,56 @@ int PhysicalMemManager::EvictPage() {
...
+  tpr[i_clock].locked = true;
+  tpr[i_clock].owner->translationTable->clearBitValid(tpr[i_clock].virtualPage);
+
+  int res = i_clock;
+
+  // copy page in swap.
+  TranslationTable* tt = tpr[res].owner->translationTable;
+  int vpn = tpr[i_clock].virtualPage;
+
+  while(tt->getBitIo(vpn)) {
```
Comment (sbihel) [inline nachos/vm/physMem.cc]: I think this doesn't work with a Sleep(). The thread will never be put back in the ready list (?). Should just use Yield().

File: nachos/vm/physMem.cc [hunk 5]
```diff
@@ -197,24 +181,56 @@ int PhysicalMemManager::EvictPage() {
...
-      g_machine->mmu->translationTable->clearBitU(tpr[i_clock_local].virtualPage);
-    } else if(!tpr[i_clock_local].locked) {
-      g_machine->mmu->translationTable->setBitU(tpr[i_clock_local].virtualPage);
-      return i_clock_local;
+    if(! tpr[i_clock].locked) {
+      tpr[i_clock].owner->translationTable->clearBitU(tpr[i_clock].virtualPage);
+    } else {
+      i++;
+    }
+    if(i >= g_cfg->NumPhysPages) {
```
Comment (sbihel) [inline nachos/vm/physMem.cc]: ?

File: nachos/kernel/scheduler.cc (hunk 1 above)
Comment (lbeziaud) [inline nachos/kernel/scheduler.cc]: You're right, I had a problem during the switch do I disabled it. We should open an issue and look into it. Note that it is fully working without deletion.

File: nachos/kernel/thread.h (hunk 2 above)
Comment (sbihel) [inline nachos/kernel/thread.h]: Isn't a local variable in the `evict` method good enough?

File: nachos/vm/physMem.cc (hunk 5 above)
Comment (lbeziaud) [inline nachos/vm/physMem.cc]: ?

File: nachos/vm/physMem.cc (hunk 4 above)
Comment (lbeziaud) [inline nachos/vm/physMem.cc]: Ah? We should definitely test this!

File: nachos/vm/physMem.cc (hunk 5 above)
Comment (sbihel) [inline nachos/vm/physMem.cc]: I don't understand this part. `i` is never modified if it is `>= g_cfg->NumPhysPages`

Comment (lbeziaud) [inline nachos/vm/physMem.cc]: We found nothing so we just wait and run all other active threads until going back to this one. The `i_clock` will change (?) so we don't touch anything.

File: nachos/kernel/thread.h (hunk 2 above)
Comment (lbeziaud) [inline nachos/kernel/thread.h]: It may be. I'm not convinced

Comment (lbeziaud) [inline nachos/kernel/thread.h]: See 69fe5c76ecbb6dc997f83ffc1a8ab36abfdb1973, seems to work fine ;)

File: nachos/vm/physMem.cc (hunk 4 above)
Comment (lbeziaud) [inline nachos/vm/physMem.cc]: Replacing `Sleep` by `Yield` in this loop doesn't work: `shell, sort` yields and error when `sort` returns: 
```
FATAL USER EXCEPTION (Thread shell, PC=0x401a18):
        *** Access to invalid or unmapped virtual address 0x405280 ***
```
A sleeping thread is still the `g_current_thread` during the next thread finding. Would that explain switching back to our sleeping thread?

File: nachos/kernel/scheduler.cc (hunk 1 above)
Comment (lbeziaud) [inline nachos/kernel/scheduler.cc]: See 6a1fb7efb8c63b4f260b60f6a7b1fc710f077850

File: nachos/vm/physMem.cc (hunk 5 above)
Comment (lbeziaud) [inline nachos/vm/physMem.cc]: See 9510fd047602664a181b3d2b78cc413eac38a078. Need to be tested more deeply.

File: nachos/vm/physMem.cc (hunk 4 above)
Comment (lbeziaud) [inline nachos/vm/physMem.cc]: That's working with the new loop, see abd352ee7dfe78a917f4745b3493048b93cba68a

**Pull request merged**